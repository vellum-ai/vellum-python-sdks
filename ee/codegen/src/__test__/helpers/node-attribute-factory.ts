import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
