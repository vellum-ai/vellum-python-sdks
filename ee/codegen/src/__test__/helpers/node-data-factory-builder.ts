import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
