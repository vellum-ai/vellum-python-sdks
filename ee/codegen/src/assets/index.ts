// This directory contains generated assets that are copied during postinstall
// and included in the npm package build. The node-definitions.json file is
// generated from the Python SDK and copied here for use in code generation.
