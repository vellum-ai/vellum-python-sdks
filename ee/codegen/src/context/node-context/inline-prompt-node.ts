import { VellumVariableType } from "vellum-ai/api";

import { BaseNodeContext } from "src/context/node-context/base";
import { PortContext } from "src/context/port-context";
import {
  NodeAttributeGenerationError,
  NodeDefinitionGenerationError,
} from "src/generators/errors";
import {
  InlinePromptNodeData,
  InlinePromptNode as InlinePromptNodeType,
  LegacyPromptNodeData,
} from "src/types/vellum";

export class InlinePromptNodeContext extends BaseNodeContext<InlinePromptNodeType> {
  baseNodeClassName = "InlinePromptNode";
  baseNodeDisplayClassName = "BaseInlinePromptNodeDisplay";

  public getNodeOutputNamesById(): Record<string, string> {
    const jsonOutput = this.nodeData.outputs?.find(
      (output) => output.type === "JSON"
    );
    const errorOutputId = this.getErrorOutputId();

    return {
      [this.nodeData.data.outputId]: "text",
      ...(errorOutputId ? { [errorOutputId]: "error" } : {}),
      [this.nodeData.data.arrayOutputId]: "results",
      ...(jsonOutput ? { [jsonOutput.id]: "json" } : {}),
    };
  }

  protected getNodeOutputTypesById(): Record<string, VellumVariableType> {
    const jsonOutput = this.nodeData.outputs?.find(
      (output) => output.type === "JSON"
    );

    return {
      [this.nodeData.data.outputId]: "STRING",
      ...(this.nodeData.data.errorOutputId
        ? { [this.nodeData.data.errorOutputId]: "ERROR" }
        : {}),
      [this.nodeData.data.arrayOutputId]: "ARRAY",
      ...(jsonOutput ? { [jsonOutput.id]: "JSON" } : {}),
    };
  }

  createPortContexts(): PortContext[] {
    return [
      new PortContext({
        workflowContext: this.workflowContext,
        nodeContext: this,
        portId: this.nodeData.data.sourceHandleId,
      }),
    ];
  }

  async buildProperties(): Promise<void> {
    // @ts-expect-error In the legacy case, we simply convert from LEGACY to INLINE on the fly after the context is initialized with the legacy node data.
    if (this.nodeData.data.variant !== "LEGACY") {
      return;
    }

    // @ts-expect-error
    const legacyNodeData: LegacyPromptNodeData = this.nodeData.data;

    const promptVersionData =
      legacyNodeData.sandboxRoutingConfig.promptVersionData;
    if (!promptVersionData) {
      throw new NodeDefinitionGenerationError(`Prompt version data not found`);
    }

    // Dynamically fetch the ML Model's name via API
    const mlModelName = await this.workflowContext
      .getMLModelNameById(promptVersionData.mlModelToWorkspaceId)
      .catch((error) => {
        this.workflowContext.addError(
          new NodeAttributeGenerationError(
            `Failed to fetch ML model name for ID ${promptVersionData.mlModelToWorkspaceId}: ${error}`,
            "WARNING"
          )
        );
        return "";
      });

    const inlinePromptNodeData: InlinePromptNodeData = {
      ...legacyNodeData,
      variant: "INLINE",
      mlModelName,
      execConfig: promptVersionData.execConfig,
    };

    this.nodeData = { ...this.nodeData, data: inlinePromptNodeData };
  }
}
