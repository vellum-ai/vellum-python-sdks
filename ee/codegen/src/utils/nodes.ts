import { VellumError } from "vellum-ai/errors";

import { WorkflowNode } from "src/types/vellum";

export function getNodeLabel(nodeData: WorkflowNode): string {
  switch (nodeData.type) {
    case "GENERIC":
      return nodeData.definition?.name ?? nodeData.label ?? "Generic Node";
    default:
      return nodeData.data.label;
  }
}

export function isUnaryOperator(operator: string): boolean {
  return operator === "null" || operator === "notNull";
}

export function isVellumErrorWithDetail(
  error: unknown
): error is VellumError & { body: { detail: string } } {
  return (
    error instanceof VellumError &&
    typeof error.body === "object" &&
    error.body !== null &&
    "detail" in error.body &&
    typeof error.body.detail === "string"
  );
}
