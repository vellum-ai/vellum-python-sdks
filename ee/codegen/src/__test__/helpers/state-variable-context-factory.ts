import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
