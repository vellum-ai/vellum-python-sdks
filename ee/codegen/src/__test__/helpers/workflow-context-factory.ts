import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
