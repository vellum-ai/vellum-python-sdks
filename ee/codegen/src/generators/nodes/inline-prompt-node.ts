import { python } from "@fern-api/python-ast";
import { AstNode } from "@fern-api/python-ast/core/AstNode";

import { OUTPUTS_CLASS_NAME } from "src/constants";
import { InlinePromptNodeContext } from "src/context/node-context/inline-prompt-node";
import { PromptTemplateBlockExcludingFunctionDefinition } from "src/generators/base-prompt-block";
import { NodeAttributeGenerationError } from "src/generators/errors";
import { FunctionDefinition } from "src/generators/function-definition";
import { BaseSingleFileNode } from "src/generators/nodes/bases/single-file-base";
import { PromptBlock } from "src/generators/prompt-block";
import { PromptParameters } from "src/generators/prompt-parameters-request";
import { WorkflowValueDescriptor } from "src/generators/workflow-value-descriptor";
import {
  InlinePromptNode as InlinePromptNodeType,
  FunctionDefinitionPromptTemplateBlock,
  InlinePromptNodeData,
  PlainTextPromptTemplateBlock,
} from "src/types/vellum";

const INPUTS_PREFIX = "prompt_inputs";

export class InlinePromptNode extends BaseSingleFileNode<
  InlinePromptNodeType,
  InlinePromptNodeContext
> {
  protected getNodeAttributeNameByNodeInputKey(nodeInputKey: string): string {
    return `${INPUTS_PREFIX}.${nodeInputKey}`;
  }

  protected getNodeClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    if (this.nodeData.data.variant !== "INLINE") {
      throw new NodeAttributeGenerationError(
        `InlinePromptNode only supports INLINE variant. Received ${this.nodeData.data.variant}`
      );
    }

    const nodeData: InlinePromptNodeData = this.nodeData.data;
    const blocksExcludingFunctionDefinition =
      nodeData.execConfig.promptTemplateBlockData.blocks.filter(
        (
          block
        ): block is Exclude<
          PromptTemplateBlockExcludingFunctionDefinition,
          PlainTextPromptTemplateBlock
        > => block.blockType !== "FUNCTION_DEFINITION"
      );

    const functionDefinitions =
      nodeData.execConfig.promptTemplateBlockData.blocks.filter(
        (block): block is FunctionDefinitionPromptTemplateBlock =>
          block.blockType === "FUNCTION_DEFINITION"
      );

    const mlModelAttr = this.nodeData.attributes?.find(
      (attr) => attr.name === "ml_model"
    );

    // If we have the node attribute for ml model use it. Otherwise default to legacy way
    // of using ml model name
    if (mlModelAttr) {
      statements.push(
        python.field({
          name: "ml_model",
          initializer: new WorkflowValueDescriptor({
            nodeContext: this.nodeContext,
            workflowContext: this.workflowContext,
            workflowValueDescriptor: mlModelAttr.value,
          }),
        })
      );
    } else {
      statements.push(
        python.field({
          name: "ml_model",
          initializer: python.TypeInstantiation.str(nodeData.mlModelName),
        })
      );
    }

    statements.push(
      python.field({
        name: "blocks",
        initializer: python.TypeInstantiation.list(
          blocksExcludingFunctionDefinition.map((block) => {
            return new PromptBlock({
              workflowContext: this.workflowContext,
              promptBlock: block,
              inputVariableNameById: Object.fromEntries(
                this.nodeData.data.execConfig.inputVariables.map(
                  (inputVariable) => [inputVariable.id, inputVariable.key]
                )
              ),
            });
          }),
          {
            endWithComma: true,
          }
        ),
      })
    );

    statements.push(
      python.field({
        name: INPUTS_PREFIX,
        initializer: python.TypeInstantiation.dict(
          Array.from(this.nodeInputsByKey.entries()).map(([key, value]) => ({
            key: python.TypeInstantiation.str(key),
            value: value,
          })),
          {
            endWithComma: true,
          }
        ),
      })
    );

    if (functionDefinitions.length > 0) {
      statements.push(
        python.field({
          name: "functions",
          initializer: python.TypeInstantiation.list(
            functionDefinitions.map(
              (functionDefinition) =>
                new FunctionDefinition({ functionDefinition })
            ),
            {
              endWithComma: true,
            }
          ),
        })
      );
    }

    statements.push(
      python.field({
        name: "parameters",
        initializer: new PromptParameters({
          promptParametersRequest: this.nodeData.data.execConfig.parameters,
        }),
      })
    );

    return statements;
  }

  protected getNodeDisplayClassBodyStatements(): AstNode[] {
    const statements: AstNode[] = [];

    statements.push(
      python.field({
        name: "label",
        initializer: python.TypeInstantiation.str(this.nodeData.data.label),
      }),
      python.field({
        name: "node_id",
        initializer: python.TypeInstantiation.uuid(this.nodeData.id),
      }),
      python.field({
        name: "output_id",
        initializer: python.TypeInstantiation.uuid(this.nodeData.data.outputId),
      }),
      python.field({
        name: "array_output_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.arrayOutputId
        ),
      }),
      python.field({
        name: "target_handle_id",
        initializer: python.TypeInstantiation.uuid(
          this.nodeData.data.targetHandleId
        ),
      })
    );

    return statements;
  }

  protected getOutputDisplay(): python.Field {
    const jsonOutput = this.nodeData.outputs?.find(
      (output) => output.type === "JSON"
    );

    const outputDisplayEntries = [
      {
        key: python.reference({
          name: this.nodeContext.nodeClassName,
          modulePath: this.nodeContext.nodeModulePath,
          attribute: [OUTPUTS_CLASS_NAME, "text"],
        }),
        value: python.instantiateClass({
          classReference: python.reference({
            name: "NodeOutputDisplay",
            modulePath:
              this.workflowContext.sdkModulePathNames
                .NODE_DISPLAY_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "id",
              value: python.TypeInstantiation.uuid(this.nodeData.data.outputId),
            }),
            python.methodArgument({
              name: "name",
              value: python.TypeInstantiation.str("text"),
            }),
          ],
        }),
      },
      {
        key: python.reference({
          name: this.nodeContext.nodeClassName,
          modulePath: this.nodeContext.nodeModulePath,
          attribute: [OUTPUTS_CLASS_NAME, "results"],
        }),
        value: python.instantiateClass({
          classReference: python.reference({
            name: "NodeOutputDisplay",
            modulePath:
              this.workflowContext.sdkModulePathNames
                .NODE_DISPLAY_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "id",
              value: python.TypeInstantiation.uuid(
                this.nodeData.data.arrayOutputId
              ),
            }),
            python.methodArgument({
              name: "name",
              value: python.TypeInstantiation.str("results"),
            }),
          ],
        }),
      },
    ];

    if (jsonOutput) {
      outputDisplayEntries.push({
        key: python.reference({
          name: this.nodeContext.nodeClassName,
          modulePath: this.nodeContext.nodeModulePath,
          attribute: [OUTPUTS_CLASS_NAME, "json"],
        }),
        value: python.instantiateClass({
          classReference: python.reference({
            name: "NodeOutputDisplay",
            modulePath:
              this.workflowContext.sdkModulePathNames
                .NODE_DISPLAY_TYPES_MODULE_PATH,
          }),
          arguments_: [
            python.methodArgument({
              name: "id",
              value: python.TypeInstantiation.uuid(jsonOutput.id),
            }),
            python.methodArgument({
              name: "name",
              value: python.TypeInstantiation.str("json"),
            }),
          ],
        }),
      });
    }

    return python.field({
      name: "output_display",
      initializer: python.TypeInstantiation.dict(outputDisplayEntries),
    });
  }

  protected getErrorOutputId(): string | undefined {
    return this.nodeData.data.errorOutputId;
  }
}
