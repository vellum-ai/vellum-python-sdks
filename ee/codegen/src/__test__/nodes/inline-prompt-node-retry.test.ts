import { Writer } from "@fern-api/python-ast/core/Writer";
import { v4 as uuidv4 } from "uuid";
import { beforeEach, describe, expect, it } from "vitest";

import { workflowContextFactory } from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import { inlinePromptNodeDataInlineVariantFactory } from "src/__test__/helpers/node-data-factories";
import { createNodeContext, WorkflowContext } from "src/context";
import { InlinePromptNodeContext } from "src/context/node-context/inline-prompt-node";
import { InlinePromptNode } from "src/generators/nodes/inline-prompt-node";

describe("InlinePromptRetryNode", () => {
  let workflowContext: WorkflowContext;
  let writer: Writer;
  let node: InlinePromptNode;

  beforeEach(() => {
    workflowContext = workflowContextFactory();
    writer = new Writer();

    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "90c6afd3-06cc-430d-aed1-35937c062531",
          key: "text",
          type: "STRING",
        },
        workflowContext,
      })
    );
  });

  describe("basic", () => {
    beforeEach(async () => {
      const nodeData = inlinePromptNodeDataInlineVariantFactory({
        blockType: "JINJA",
        adornments: [
          {
            id: "cc79c784-d936-44c2-a811-b86a53e6ff68",
            label: "RetryNodeLabel",
            base: {
              name: "RetryNode",
              module: [
                "vellum",
                "workflows",
                "nodes",
                "core",
                "retry_node",
                "node",
              ],
            },
            attributes: [
              {
                id: uuidv4(),
                name: "max_attempts",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 3,
                  },
                },
              },
            ],
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as InlinePromptNodeContext;

      node = new InlinePromptNode({
        workflowContext,
        nodeContext,
      });
    });

    it(`getNodeFile`, async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
    it(`getNodeDisplayFile`, async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
  describe("basic retry adornment with delay", () => {
    let node: InlinePromptNode;

    beforeEach(async () => {
      const nodeData = inlinePromptNodeDataInlineVariantFactory({
        blockType: "JINJA",
        adornments: [
          {
            id: "2076aea8-be38-4ff1-8c68-cb853e352d66",
            label: "RetryNodeLabel",
            base: {
              name: "RetryNode",
              module: [
                "vellum",
                "workflows",
                "nodes",
                "core",
                "retry_node",
                "node",
              ],
            },
            attributes: [
              {
                id: uuidv4(),
                name: "max_attempts",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 3,
                  },
                },
              },
              {
                id: uuidv4(),
                name: "delay",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 2,
                  },
                },
              },
            ],
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as InlinePromptNodeContext;

      node = new InlinePromptNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });

  describe("basic retry adornment and try adornment", () => {
    let node: InlinePromptNode;
    const ERROR_OUTPUT_ID = "e7a1fbea-f5a7-4b31-a9ff-0d26c3de021f";

    beforeEach(async () => {
      const nodeData = inlinePromptNodeDataInlineVariantFactory({
        blockType: "JINJA",
        errorOutputId: ERROR_OUTPUT_ID,
        adornments: [
          {
            id: "e5de8d57-ae0d-4a4a-afb3-eb4cd6bdb0ac",
            label: "RetryNodeLabel",
            base: {
              name: "RetryNode",
              module: [
                "vellum",
                "workflows",
                "nodes",
                "core",
                "retry_node",
                "node",
              ],
            },
            attributes: [
              {
                id: uuidv4(),
                name: "max_attempts",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 3,
                  },
                },
              },
              {
                id: uuidv4(),
                name: "delay",
                value: {
                  type: "CONSTANT_VALUE",
                  value: {
                    type: "NUMBER",
                    value: 2,
                  },
                },
              },
            ],
          },
        ],
      });

      const nodeContext = (await createNodeContext({
        workflowContext,
        nodeData,
      })) as InlinePromptNodeContext;

      node = new InlinePromptNode({
        workflowContext,
        nodeContext,
      });
    });

    it("getNodeFile", async () => {
      node.getNodeFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });

    it("getNodeDisplayFile", async () => {
      node.getNodeDisplayFile().write(writer);
      expect(await writer.toStringFormatted()).toMatchSnapshot();
    });
  });
});
