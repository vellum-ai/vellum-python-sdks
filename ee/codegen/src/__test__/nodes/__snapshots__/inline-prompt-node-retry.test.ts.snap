// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InlinePromptRetryNode > basic > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.nodes import (
    BaseInlinePromptNodeDisplay,
    BaseRetryNodeDisplay,
)
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)

from ...nodes.prompt_node import PromptNode


@BaseRetryNodeDisplay.wrap(node_id=UUID("cc79c784-d936-44c2-a811-b86a53e6ff68"))
class PromptNodeDisplay(BaseInlinePromptNodeDisplay[PromptNode]):
    label = "Prompt Node"
    node_id = UUID("7e09927b-6d6f-4829-92c9-54e66bdcaf80")
    output_id = UUID("2d4f1826-de75-499a-8f84-0a690c8136ad")
    array_output_id = UUID("771c6fba-5b4a-4092-9d52-693242d7b92c")
    target_handle_id = UUID("3feb7e71-ec63-4d58-82ba-c3df829a2948")
    node_input_ids_by_name = {
        "prompt_inputs.text": UUID("7b8af68b-cf60-4fca-9c57-868042b5b616")
    }
    output_display = {
        PromptNode.Outputs.text: NodeOutputDisplay(
            id=UUID("2d4f1826-de75-499a-8f84-0a690c8136ad"), name="text"
        ),
        PromptNode.Outputs.results: NodeOutputDisplay(
            id=UUID("771c6fba-5b4a-4092-9d52-693242d7b92c"), name="results"
        ),
    }
    port_displays = {
        PromptNode.Ports.default: PortDisplayOverrides(
            id=UUID("dd8397b1-5a41-4fa0-8c24-e5dffee4fb98")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`InlinePromptRetryNode > basic > getNodeFile 1`] = `
"from vellum import JinjaPromptBlock, PromptParameters
from vellum.workflows.nodes.core.retry_node.node import RetryNode
from vellum.workflows.nodes.displayable import InlinePromptNode

from ..inputs import Inputs


@RetryNode.wrap(max_attempts=3)
class PromptNode(InlinePromptNode):
    ml_model = "gpt-4o-mini"
    blocks = [
        JinjaPromptBlock(template="""Summarize what this means {{ INPUT_VARIABLE }}"""),
    ]
    prompt_inputs = {
        "text": Inputs.text,
    }
    parameters = PromptParameters(
        stop=[],
        temperature=0,
        max_tokens=1000,
        top_p=1,
        top_k=0,
        frequency_penalty=0,
        presence_penalty=0,
        logit_bias={},
        custom_parameters={},
    )
"
`;

exports[`InlinePromptRetryNode > basic retry adornment and try adornment > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.nodes import (
    BaseInlinePromptNodeDisplay,
    BaseRetryNodeDisplay,
    BaseTryNodeDisplay,
)
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)

from ...nodes.prompt_node import PromptNode


@BaseTryNodeDisplay.wrap(
    node_id=UUID("e7a1fbea-f5a7-4b31-a9ff-0d26c3de021f"),
    error_output_id=UUID("e7a1fbea-f5a7-4b31-a9ff-0d26c3de021f"),
)
@BaseRetryNodeDisplay.wrap(node_id=UUID("e5de8d57-ae0d-4a4a-afb3-eb4cd6bdb0ac"))
class PromptNodeDisplay(BaseInlinePromptNodeDisplay[PromptNode]):
    label = "Prompt Node"
    node_id = UUID("7e09927b-6d6f-4829-92c9-54e66bdcaf80")
    output_id = UUID("2d4f1826-de75-499a-8f84-0a690c8136ad")
    array_output_id = UUID("771c6fba-5b4a-4092-9d52-693242d7b92c")
    target_handle_id = UUID("3feb7e71-ec63-4d58-82ba-c3df829a2948")
    node_input_ids_by_name = {
        "prompt_inputs.text": UUID("7b8af68b-cf60-4fca-9c57-868042b5b616")
    }
    output_display = {
        PromptNode.Outputs.text: NodeOutputDisplay(
            id=UUID("2d4f1826-de75-499a-8f84-0a690c8136ad"), name="text"
        ),
        PromptNode.Outputs.results: NodeOutputDisplay(
            id=UUID("771c6fba-5b4a-4092-9d52-693242d7b92c"), name="results"
        ),
    }
    port_displays = {
        PromptNode.Ports.default: PortDisplayOverrides(
            id=UUID("dd8397b1-5a41-4fa0-8c24-e5dffee4fb98")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`InlinePromptRetryNode > basic retry adornment and try adornment > getNodeFile 1`] = `
"from vellum import JinjaPromptBlock, PromptParameters
from vellum.workflows.nodes.core import TryNode
from vellum.workflows.nodes.core.retry_node.node import RetryNode
from vellum.workflows.nodes.displayable import InlinePromptNode

from ..inputs import Inputs


@TryNode.wrap()
@RetryNode.wrap(max_attempts=3, delay=2)
class PromptNode(InlinePromptNode):
    ml_model = "gpt-4o-mini"
    blocks = [
        JinjaPromptBlock(template="""Summarize what this means {{ INPUT_VARIABLE }}"""),
    ]
    prompt_inputs = {
        "text": Inputs.text,
    }
    parameters = PromptParameters(
        stop=[],
        temperature=0,
        max_tokens=1000,
        top_p=1,
        top_k=0,
        frequency_penalty=0,
        presence_penalty=0,
        logit_bias={},
        custom_parameters={},
    )
"
`;

exports[`InlinePromptRetryNode > basic retry adornment with delay > getNodeDisplayFile 1`] = `
"from uuid import UUID

from vellum_ee.workflows.display.editor import NodeDisplayData, NodeDisplayPosition
from vellum_ee.workflows.display.nodes import (
    BaseInlinePromptNodeDisplay,
    BaseRetryNodeDisplay,
)
from vellum_ee.workflows.display.nodes.types import (
    NodeOutputDisplay,
    PortDisplayOverrides,
)

from ...nodes.prompt_node import PromptNode


@BaseRetryNodeDisplay.wrap(node_id=UUID("2076aea8-be38-4ff1-8c68-cb853e352d66"))
class PromptNodeDisplay(BaseInlinePromptNodeDisplay[PromptNode]):
    label = "Prompt Node"
    node_id = UUID("7e09927b-6d6f-4829-92c9-54e66bdcaf80")
    output_id = UUID("2d4f1826-de75-499a-8f84-0a690c8136ad")
    array_output_id = UUID("771c6fba-5b4a-4092-9d52-693242d7b92c")
    target_handle_id = UUID("3feb7e71-ec63-4d58-82ba-c3df829a2948")
    node_input_ids_by_name = {
        "prompt_inputs.text": UUID("7b8af68b-cf60-4fca-9c57-868042b5b616")
    }
    output_display = {
        PromptNode.Outputs.text: NodeOutputDisplay(
            id=UUID("2d4f1826-de75-499a-8f84-0a690c8136ad"), name="text"
        ),
        PromptNode.Outputs.results: NodeOutputDisplay(
            id=UUID("771c6fba-5b4a-4092-9d52-693242d7b92c"), name="results"
        ),
    }
    port_displays = {
        PromptNode.Ports.default: PortDisplayOverrides(
            id=UUID("dd8397b1-5a41-4fa0-8c24-e5dffee4fb98")
        )
    }
    display_data = NodeDisplayData(
        position=NodeDisplayPosition(x=0, y=0), width=None, height=None
    )
"
`;

exports[`InlinePromptRetryNode > basic retry adornment with delay > getNodeFile 1`] = `
"from vellum import JinjaPromptBlock, PromptParameters
from vellum.workflows.nodes.core.retry_node.node import RetryNode
from vellum.workflows.nodes.displayable import InlinePromptNode

from ..inputs import Inputs


@RetryNode.wrap(max_attempts=3, delay=2)
class PromptNode(InlinePromptNode):
    ml_model = "gpt-4o-mini"
    blocks = [
        JinjaPromptBlock(template="""Summarize what this means {{ INPUT_VARIABLE }}"""),
    ]
    prompt_inputs = {
        "text": Inputs.text,
    }
    parameters = PromptParameters(
        stop=[],
        temperature=0,
        max_tokens=1000,
        top_p=1,
        top_k=0,
        frequency_penalty=0,
        presence_penalty=0,
        logit_bias={},
        custom_parameters={},
    )
"
`;
